c++/src/kj/compat/tls.c++

```diff
@@ -419,20 +419,26 @@ private:
 
 class TlsConnectionReceiver final: public ConnectionReceiver, public TaskSet::ErrorHandler {
 public:
-  TlsConnectionReceiver(TlsContext &tls, Own<ConnectionReceiver> inner)
+  TlsConnectionReceiver(
+      TlsContext &tls, Own<ConnectionReceiver> inner, kj::Maybe<ErrorHandler&> errorHandler)
       : tls(tls), inner(kj::mv(inner)),
         acceptLoopTask(acceptLoop().eagerlyEvaluate([this](Exception &&e) {
           onAcceptFailure(kj::mv(e));
         })),
-        tasks(*this) {}
+        tasks(errorHandler.orDefault(*this)) {}
 
   void taskFailed(Exception&& e) override {
-    // TODO(someday): SSL connection failures may be a fact of normal operation but they may also
-    // be important diagnostic information. We should allow for an error handler to be passed in so
-    // that network issues that affect TLS can be more discoverable from the server side.
-    if (e.getType() != Exception::Type::DISCONNECTED) {
-      KJ_LOG(ERROR, "error accepting tls connection", kj::mv(e));
-    }
+    // Note that if an ErrorHandler was provided to the ctor, this will not be used.
+    switch (e.getType()) {
+      case Exception::Type::DISCONNECTED:
+      case Exception::Type::OVERLOADED: {
```

Comment:
Ideally it would be better if the timeout code wouldn't produce an OVERLOADED exception in the first place, rather than have this code trying to detect timeout errors by catching this error type.

The OVERLOADED error is a side-effect of using `timer.timeoutAfter()`, but it's not actually even the right error here, as this timeout doesn't actually suggest anything is overloaded. The reason `timeoutAfter()` chooses OVERLOADED as its exception type is because it's meant to be used on the client side when talking to a remote server -- if the server doesn't respond, we guess it's overloaded. But here we're on the server side and waiting for a client, and the assumption doesn't really hold.

The right way to get more control over how timeouts manifest is to use `promise.exclusiveJoin()` with `timer.afterDelay(timeout)`. This is e.g. how KJ's HTTP server implements timeouts. You could, for example, treat timeouts as DISCONNECTED with something like:

```
promise = promise.exclusiveJoin(
    timer.afterDelay(timeout).then([]() -> kj::Promise<void> {
  return KJ_EXCEPTION(DISCONNECTED, "client timed out during TLS handshake");
}))
```

Or, if we didn't want to think of timeouts as errors at all (my preference), but we still wanted to allow a custom error handler to gather metrics about them, we'd do something like:

```
promise = promise.exclusiveJoin(
    timer.afterDelay(timeout).then([this]() {
  errorHandler.notifyTimeout();
}))
```

This requires a dedicated error handler interface, though.

In retrospect I sort of regret adding `timer.timeoutAfter()`, or at least I wish I'd made it require the caller to specify the exception to use on error. I think the default of OVERLOADED has led to a lot of confusion about exception types...

Reply:
I was wondering if you'd prefer the afterDelay form. I think I do as well. I'll switch it over.

If you're okay with it, I'll skip the custom error handler.